Incremental decommit

ch1bo opened the pull request:
This PR adds "incremental decommits" to the Hydra Head protocol, which allows users to take funds out of an open Head.

- New API endpoint `/decommit` which accepts a "Decommit transaction", that spends some UTxO and whatever outputs it produces will be made available on the L1. This can be also done through a new `Decommit` client input and new server outputs `DecommitRequested`, `DecommitApproved` and `DecommitFinalized`, as well as `DecommitInvalid` to inform about status of the decommit.

- Decommits are first approved in a snapshot on L2 via a new network message `ReqDec`, before a new `decrementTx` can be posted and observed on-chain.

- Only one decommit can be processed at a given time.

- Update documentation and added how-to about how to use this.

- Acknowledged specification changes by "clearing" of $\textcolor{red}{\\red}$ areas covered by this implementation in the specification.

- End-to-end test covering the main scenario of decommitting funds.

- Added mutation tests for Decrement, Close and Contest to cover all on-chain-verification changes.

- Enhanced `TxTrace` tests to test decrements with various snapshots and their interaction with close/contest and fanout of a head.

---

* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Haddocks updated
* [x] New TODOs explained hereafter

![image](https://github.com/user-attachments/assets/eed47f06-d519-42cb-a897-98397066fdd9)

  - Two FIXMEs covered by #1524
  - TODO in HeadLogic coverd by #1502
  - TODO in tx-cost how we could improve the benchmark output (not crucial)
  - TODO in head logic about rollbacks .. actually something we need to consider with #199 too

ch1bo committed:
TxTrace: increase number of positive actions

This should provide "happier" paths, while still trying unhappy paths
~20% of the time.

ch1bo committed:
TxTrace: Model Augmented snapshots

This results in snapshots with non-empty utxoToDecommit used in the
decrementTx and currently fails to build a valid tx (maybe incomplete model)

ch1bo committed:
TxTrace: identify expected failing fanout

ch1bo committed:
TxTrace: Refactor into a short-cutting PostconditionM'

ch1bo committed:
TxTrace: keep construction errors in TxResult

Also fix the rebase compilation failures

v0d1ch committed:
Fix the failing close with initial snapshot

v0d1ch committed:
TxTrace: Add decommit utxo to the snapshot

This produces errors with negative values in the output because we
negate the head output value by the decommit amount.

Also fix the rebase compilation failures.

v0d1ch committed:
TxTrace: re-sign the decommit snapshot

After we add utxoToDecommit to existing snapshot we need to re-sign it.

v0d1ch committed:
TxTrace: Do not remove decommit outputs

We produce random snapshots so we don't need to remove the decommitted
outputs from the current UTxO set.

v0d1ch committed:
Fix golden files

v0d1ch committed:
Simplify getting the snapshotNumber in closeTx

v0d1ch committed:
TxTrace: Produce decommit snapshots directly

When doing a decommit we need to make sure that the decommit utxo from
the snapshot is added the the head UTxO, otherwise we get
unbalanced/negative values. This leads to problems with closed/fanout
utxo hashes `H24`.

v0d1ch committed:
TxTrace: Add decommit value to the head UTxO

This prevents negative values in the head outputs since our snapshots
are generated randomly and are not connected to the actual head utxo.

v0d1ch committed:
Add utxoToDecommit to FanoutTx

This is needed to determine the proper hash of the outputs in the checkFanout
validator.

ch1bo committed:
TxTrace: Not add value to make decrement pass

Modifying the value defeats the purpose of model-based testing and
potential assertions of the real world matching our model.

v0d1ch committed:
TxTrace: Remove Augmented Snapshot and alter the model

Introduce ModelUTxO and utxoInHead to the model.

Also remove commented out code related to fanout.

v0d1ch committed:
TxTrace: Add modelUTxO to the snapshot

v0d1ch committed:
TxTrace: Produce a valid model without accounting for decrements

v0d1ch committed:
TxTrace: Track decommit UTxO in ModelUTxO

I think I found an easy way to pick something to decommit. I always
generate utxo plus utxo to decommit. When producing the actuall real
snapshot I can just choose to set to mempty one of the fields in the
ModelSnapshot (snapshotUTxO or decommitUTxO) which should give me a
proper subset of something to decommit without any extra changes.

Now is the time to play with constraints in the
arbitraryAction/precondition to make sure we always generate something
sensible.

v0d1ch committed:
TxTrace: Remove utxoTeDecrement from the Model

v0d1ch committed:
TxTrace: Generate Decrement and Contest actions using existing head utxo

v0d1ch committed:
TxTrace: If there is no utxo in the snapshot don't populate
utxoToDecommit also

v0d1ch committed:
TxTrace: precondition checks

v0d1ch committed:
Rebase: fix compilation

v0d1ch committed:
TxTrace: Remove validFailingActions and revert conditions in the
precondition.

We wanted to start fresh. With these changes all we see is H24 in the
fanout validator so that is what we will fix next.

v0d1ch committed:
TxTrace: Generate separatelly values for utxo and utxo to decommit

v0d1ch committed:
Fix bug related to intial utxo hash for the decommit values

v0d1ch committed:
TxTrace: Re-add validFailingAction and introduce more arbitrary values
in the generator

ch1bo committed:
Refactor performTx and TxResult + some documentation

This should make it clear how this is currently used in between perform
and postcondition.

ch1bo committed:
Remove some preconditions to see more radical failures

ch1bo committed:
Add more information on expected validation failures

ch1bo committed:
Fix starting real world utxo to included "A" equivalent

ch1bo committed:
Exercise more negative actions (by just trying all which don't fulfill
the precondition)

ffakenz committed:
Fix how the model snapshot is interpreted into a confirmed in case of the initial

ffakenz committed:
Discard decrements that attempts to take more value than it is in the head

We changed the decommitSnapshot that interprets the model into the real world to
expose the TxBodyOutputNegative Bug.
Because its an invalid action we dont want to deal with it, so we decided to discard it.

ffakenz committed:
Increase frequency of negative Decrement actions being produced

And we also make sure that when we produce positive Decrements its
because there is some utxo to decrement and not empty.

ffakenz committed:
Add more interested cases to cover

ffakenz committed:
Discard Stops

ffakenz committed:
Add realWorldModelUTxO preserves addition property

Enhance the model to include the utxo balance while preserving its identity.

v0d1ch committed:
TxTrace: Alter the ModelUTxO to use Map instead of Set

Also make sure to check the decommit value is not zero since we want to
assert we are actually decommitting something.

v0d1ch committed:
TxTrace: initially start with some funds

ch1bo committed:
Slightly expand the ModelUTxO alphabet and exercise more decommits

This is very similar as before, but generates decommits with subsets of
what is included in the head and only discards decrements which would
run into non-gracefully handled negative balance situations.

ffakenz committed:
Lax the precondition for Decrement

As it was expecting only submaps to be valid.
This was preventing from taking partial values for a given key to decommit.

Also fixed utxoInHead after nextState:
* We were removing all the balance for a given key to decommit,
regardless of the value to be decommited.

v0d1ch committed:
WIP

v0d1ch committed:
Close Mutation: Set utxoToDecommit to some value

Make sure to exercise the path where we change the utxoToDecommit hash
in the close datum in order to see invalid snapshot signature.

v0d1ch committed:
Fanout Mutation: Set utxoToDecommit in the fanout snapshot

Exercise the path where the fanout snapshot is created with something to
decommit and then alter the outputs to add one more so the hash doesn't
match with the fanout redeemer.

v0d1ch committed:
Fanout Mutation: Alter the fanout redeemer

If we alter the Fanout redeemer this will trigger the
FannedOutUtxoHashNotEqualToClosedUtxoHash since we hash the outputs
(both normal and decommit ones) and compare with the close datum.

v0d1ch committed:
Remove PartySignatureVerificationFailed since it is not used

This change also reduces the size of or script which is needed because
of decommits which made it too large.

v0d1ch committed:
Contest Mutation: Add utxoToDecommit to the contest snapshot

Exercise altering the hashes of utxo and utxoToDecommit to trigger the
contestation signature errors `SignatureVerificationFailed`.

v0d1ch committed:
Optimize datum extraction to reduce execution costs

ffakenz committed:
Reject decrements with empty utxo to decommit

Adapt tx-traces to not generate decrements if there is no utxo in the head.
Also added detail comments to Decrement mutation tests.

v0d1ch committed:
Improve BehaviorSpec

Make sure utxoToDecommit is not fanned out and close with decommit in
place will fanout both utxos present.

v0d1ch committed:
TxSpec: Mutate working decrement snapshots

Idea is to mutate different snapshot fields in order to produce
expected errors in the validator. I think this way we should be able to
test the decrement, close and fanout using different (valid or invalid)
snapshots and get to assert what kind of errors we expect.

v0d1ch committed:
TxSpec: Assert valid sequence of actions work

Note: This is using a constant snapshot. I think it is better to go back
to modeling expected errors when altering snapshots than this tx level
type testing.

v0d1ch committed:
TxSpec: WIP

v0d1ch committed:
TxSpec: Mutate snapshots and utxo to produce expected errors

ffakenz committed:
Enhance interesting transaction traces

Add coverage for transactions with many decrements

ffakenz committed:
Dry how to generate model snapshots

ffakenz committed:
Minor dry on balanceUTxOInHead

v0d1ch committed:
TxSpec: Make sequence of valid txs work

v0d1ch committed:
TxSpec: Re-enable all error cases

v0d1ch committed:
TxSpec: No need for property tests and suchThat

We generate perfect ModelSnapshot anyway and then alter it (or alter the
UTxO) to get expected failures so we don't depend on anything arbitrary
here.

v0d1ch committed:
TxSpec: Remove labels as they are not very useful

v0d1ch committed:
TxSpec: Simplify and do property checks in individual txs

v0d1ch committed:
TxSpec: Refactor actions

locallycompact committed:
Distinguish between missing snapshot and null snapshot errors

ffakenz committed:
Ignore decrements without something to decommit

ffakenz committed:
Remove keys with 0 values from someUTxOToDecrement

ffakenz committed:
Keep track of decommitUTxOInHead

So we can preserve that valid fanouts are built from valid close results

ffakenz committed:
Apply same ignore logic to contest

ffakenz committed:
Enhance hasManyDecommits rule

To actually verify there are many decommits trying to take value out of the head

ffakenz committed:
Remove the has many decommits coverage for close and fanout

As we were getting 0% generaged.
Also reduce the expected coverage for close and fanout with something to decommit.

v0d1ch committed:
TxSpec: Test multiple chained actions with altered snapshots

v0d1ch committed:
TxSpec: Build combinators

No pleasant way of testing negative actions in QuickCheck it seems

v0d1ch committed:
TxSpec: Use Bool instead of Property

Since we can't expect Property in some cases to return false
(expectedFailure doesn't work with conjuctions) I used [Bool] instead.
It is a bit lame since we can't collect errors but on the other hand the
tracing is easy.

v0d1ch committed:
Fix LoggingSpec test

v0d1ch committed:
Fix StateSpec test by not generating empty commits

If we always need something to decrement the easiest way to do that is
to not generate empty commits.

v0d1ch committed:
Fix golden files

v0d1ch committed:
Plutus scripts

v0d1ch committed:
Format

v0d1ch committed:
Remove pretty-simple

v0d1ch committed:
Make splitUTxO pure

v0d1ch committed:
Use utxoToDecommit in a snapshot in DirectChainSpec

v0d1ch committed:
Delete quickstart

v0d1ch committed:
Improve splitUTxO

v0d1ch committed:
Add Decrement transaction to tx-cost benchmark

Also use utxoToDecommit in other tx snapshots

v0d1ch committed:
Use something to decommit in the snapshot DirectChainSpec

v0d1ch committed:
Add splitUTxO property test

v0d1ch committed:
Arbitrary Snapshot should have something to decommit

Propagate utxo to decommit to close/contest/fanout in State

v0d1ch committed:
Document Snapshot CDDL

Spec: Mainly remove incremental commit mentioning
Also add some decommit parts
Spec: Try to tacke fanout to a point
Explain better decommit command/message

v0d1ch committed:
Add changelog entry

ffakenz committed:
Move decommit tutorial to how-to section

v0d1ch committed:
Expose a bug when decommitting

ffakenz committed:
Fix how-to tutorial: do not sign decommit-tx with alice key

v0d1ch committed:
Add Version: Introduce version to the HeadState

v0d1ch committed:
Add Version: Introduce version to Snapshot

v0d1ch committed:
Add Version: Require version in local state and confirmed snapshot to
match

v0d1ch committed:
Add Version: Start with version 0 and only bump on observed decommit

v0d1ch committed:
Add Version: Add missing version in Snapshot toJSON and toCBOR

Also improve on docs for Snapshot version

v0d1ch committed:
Add Version: Default to version 0 in tests to pass validation

v0d1ch committed:
Add Version: Remove required version match check

Both come from local state so this check is redundant

v0d1ch committed:
Add Version: Model the snapshot version into TxTrace

v0d1ch committed:
Add Version: Bump execution costs for now

v0d1ch committed:
Add Version: Add off-chain version to CloseTx

v0d1ch committed:
Add Version: Add closeUTxOToDecommit to CloseTx

v0d1ch committed:
Add Version: Add utxoToDecommitHash to Close redeemer

Alter off-chain code to change the close datum depending on
versions and on-chain code to check the proper hash, either from
the closed datum or closed redeemer.

v0d1ch committed:
Add Version: Add version to the Closed off-chain state

Decide how to populate utxoToDecommitHash on FanoutTx depending
on Closed version. This enables us not alter on-chain code and
forward the expected utxo to decommit outputs depending on the version.

v0d1ch committed:
Add Version: Mutate number of decommit outputs on Fanout

v0d1ch committed:
Add Version: If we never update the version all transitions should still
work

I couldn't model the version changes the way I wanted here so I opted
out to not do it right away.

v0d1ch committed:
Add Version: Optimize head contract using PlutusTx.Plugin
conservative-optimisation

ffakenz committed:
Introduced newtype for snapshot version instead of plain natural.

This maintains consistency with snapshot number.

ffakenz committed:
Move snapshot version into coordinated head state in open state

This respects the spec.

ffakenz committed:
Minor fix on how the balance utxo occurs in TxTraceSpec

ffakenz committed:
Remove unnecessary long comment from Head script

v0d1ch committed:
Add Version: Introduce optimisations to other contracts

ffakenz committed:
Introduce new HeadError for IncorrectVersion.
This is used to check version inits with 0 in CollectCom

ffakenz committed:
Add mutation check over decrement to very it increases the version number

This change forced me to draft a refactor on the close validator
to verify the snapshot signature depending on the close type.

v0d1ch committed:
Add Version: Make version check in close more readable

v0d1ch committed:
Add Version: Fix the decommit mutation for lucky seed

v0d1ch committed:
Add Version: Make Close isHealthy check pass

ffakenz committed:
Add minor doc comments

ffakenz committed:
Propose new head logic specs

v0d1ch committed:
Fix logging spec errors

ffakenz committed:
Add e2e spec to check: can close with a decommit in flight

ffakenz committed:
Reorder check close in head script so mutations fail for the right reason

ffakenz committed:
Minor refactor: split close mutations according to their type

v0d1ch committed:
Refactor CloseCurrent more - All Close mutations pass now

ffakenz committed:
Draft close outdated mutation specs

ffakenz committed:
Minor refactor on close mutations

Reorder again and remove redundant functions

v0d1ch committed:
Fix checkClose for the version check

We did - 1 which in case of verson 0 produces a negative number
which in turn produces invalid signature check.

Also fix the e2e scenario where we used the wrong version.

v0d1ch committed:
Fix StateSpec for decommit and close

Basically alter the versions so they are what we expect.

v0d1ch committed:
Fix TxSpec for chained transactions

v0d1ch committed:
Fix "can only contest once" with proper version

v0d1ch committed:
Fix the golden files

v0d1ch committed:
Improve CloseOutdated mutations

v0d1ch committed:
CloseOutdated: Change decommit hash in the redeemer

v0d1ch committed:
CloseOutdated: Change the signatures in the redeemer

ffakenz committed:
Refactor close current to follow the spec more

Also complete mutations cases for H34 (signature verification failed)
Keep the outdated mutations aligned with current just for consistency:
- we are still missing to change the input head datum so that it reproduces the outdated case

WARN: Although this commit breaks the e2e spec we believe this is a good step to move forward.

v0d1ch committed:
Remove unrelevant mutation from CloseCurrent

ffakenz committed:
Add check to verify the version is preserved btw open and close states

v0d1ch committed:
Gain some sanity

ffakenz committed:
Finish mutation test for outdated close

TODO left:
> Check the snapshot version is preserved from last open state.

ffakenz committed:
Introduce contest current mutations

ffakenz committed:
Add contest redeemer type

We followed the same logic applied for close.

v0d1ch committed:
Add some notes to explain outstandings in head validator

ffakenz committed:
Add snapshot version to ReqSn

ffakenz committed:
Minor fix to direct chain spec

The close tx utxo to decommit should match whatever is in the l2
or FailedToConstructCloseTx will be thrown.

ffakenz committed:
Align on reqDec to spec

Calculate missing step: L̂ <- L̂ \ inputs(tx)
and its been addded to DecommitRecorded.

ffakenz committed:
Align on reqSn to spec

Add missing step: require v = vˆ
This introduced a new RequirementFailure: ReqSvNumberInvalid

ffakenz committed:
Align on reqSn to spec

Add missing step: require v = vˆ
This introduced a new RequirementFailure: ReqSvNumberInvalid

ffakenz committed:
Align reqSn with spec

Only minor reordering of steps to follow the spec.

v0d1ch committed:
Add "OffChain" version to TxTrace

Also remove contest check for version since it should not work this way.

v0d1ch committed:
TxTrace: Bump snapshot version after successful decrement

ffakenz committed:
Align on decrementTx from chain with spec

Note: should we get Uω from observation instead of local state?

v0d1ch committed:
Fix initial snapshot close in mock chain.

This actually fixes our Model spec

ffakenz committed:
Spot gaps btw onOpenNetworkAckSn and spec

ffakenz committed:
Add spec comments onOpenClientClose according to latest spec changes

v0d1ch committed:
Spec: revert unnecessary change in overview.tex

locallycompact committed:
Refactor common healthy fixtures

v0d1ch committed:
Match on snapshot number being zero to detemine if we are in initial
redeemer.

v0d1ch committed:
Make CloseInitial tx healthy again

v0d1ch committed:
Make CloseOutdated work again

v0d1ch committed:
Format

ffakenz committed:
Align offchain protocol to spec and documment missing gaps

v0d1ch committed:
Spec: use m' instead of n for the fanout redeemer outputs

v0d1ch committed:
Fanout mutations

We splitted the fanout errors into two types:
FannedOutUtxoHashNotEqualToClosedUtxoHash and
FannedOutUtxoHashNotEqualToClosedUtxoHashToDecommit

v0d1ch committed:
Fix the logs schema errors

v0d1ch committed:
Fix TxTraceSpec

v0d1ch committed:
Fix the scripts

v0d1ch committed:
Spec: Update contestTx.pdf

ffakenz committed:
Align on close and context tx to spec and documment missing gaps

v0d1ch committed:
Spec: Update fanoutTx.pdf

v0d1ch committed:
Spec: Small fix for the number of decommit outputs in fanout tx

v0d1ch committed:
E2E: Add one more decommit to test multiple decommits

v0d1ch committed:
Update submodule conway.genesis files

v0d1ch committed:
Improve decommit tutorial instructions

v0d1ch commented on hydra-plutus/src/Hydra/Contract/HeadState.hs:
```diff
@@ -30,32 +32,55 @@ data State
       , utxoHash :: Hash
+      -- ^ Spec: η
       , headId :: CurrencySymbol
+      , snapshotNumber :: SnapshotNumber
```
Spec gap: snapshot number should not be in the datum.

v0d1ch commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -155,11 +157,9 @@ checkCollectCom ctx@ScriptContext{scriptContextTxInfo = txInfo} (contestationPer
-        && headId' == headId
+  mustInitVersion =
+    traceIfFalse $(errorCode IncorrectVersion) $
+      version' == sn && sn == 0
```
checking that version == 0 should be enough here

v0d1ch commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -275,15 +258,47 @@ genCollectComMutation (tx, _utxo) =
     st -> st
+
+  mutateVersion = do
+    mutatedVersion <- arbitrary `suchThat` (> 0)
```
anything other than zero should trigger the validator

v0d1ch commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -219,13 +209,61 @@ checkCollectCom ctx@ScriptContext{scriptContextTxInfo = txInfo} (contestationPer
+    && mustBeSignedByParticipant ctx prevHeadId
+    && mustDecreaseValue
+ where
+  checkSnapshot =
```
checkSnapshot should not be needed.

v0d1ch commented on hydra-plutus/src/Hydra/Contract/HeadState.hs:
```diff
@@ -30,32 +32,55 @@ data State
 
 PlutusTx.unstableMakeIsData ''State
 
+data Version
```
Separate version redeemer for the contest and also rename to maybe Close/ContestAction ?

v0d1ch replied:
Keep the needed fields for on-chain checks in the redeemer type.

v0d1ch commented on hydra-plutus/src/Hydra/Contract/HeadState.hs:
```diff
@@ -30,32 +32,55 @@ data State
       , utxoHash :: Hash
+      -- ^ Spec: η
+      , utxoToDecommitHash :: Hash
+      -- ^ Spec: ηω
```
Rename and make sure we use different naming for eta omega in the snapshot and in the on-chain parts.

v0d1ch commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -272,27 +320,36 @@ checkClose ctx parties initialUtxoHash sig cperiod headPolicyId =
+      CurrentVersion ->
+        (decommitHash, outputVersion)
+      OutdatedVersion ->
+        -- TODO: missing to check -> η∆′ = ⊥ Is this really needed? Signature
```
Valid todo. Implement it.

ch1bo commented:
Taking over the last sub-task of going through the implementation + spec and "clearing out red areas" by aligning the implementation with the specification.

ch1bo commented on docs/core-concepts/index.md:
````diff
@@ -10,3 +10,44 @@ import {useDocsSidebar} from '@docusaurus/theme-common/internal';
 <DocCardList items={useDocsSidebar().items.filter(({ docId }) => docId != "index")}/>
 ```
+
+### Incremental Decommits
````
Move to a dedicated documentation page?

ch1bo commented on CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ changes.
 
 ## [0.18.0] - Unreleased
 
+- Add capability to decommit or take some funds out of a running Head.
```
Expand

ch1bo commented on docs/docs/how-to/incremental-decommit.md:
```diff
@@ -0,0 +1,133 @@
+   | jq "select(.tag == \"Greetings\") \
+   | .snapshotUtxo \
+   | with_entries(select(.value.address == \"$(cat alice-funds.addr)\"))" \
+   > utxo.json
```
Could: Use `GET /snapshot/utxo` instead

ch1bo commented on docs/docs/how-to/incremental-decommit.md:
```diff
@@ -0,0 +1,133 @@
+---
+
+# Decommit funds
+> This example assumes you have the hydra-node repo at your disposal together with `hydra-node`, `hydra-tui`, `cardano-cli` and `curl` binaries and that cardano-node is running on local devnet.
```
Too restrictive?

ch1bo commented on hydra-prelude/src/Hydra/Prelude.hs:
```diff
@@ -249,3 +255,11 @@ withFile fp mode action =
+
+-- | Like 'spy' but prefixed with a label.
+spy' :: Show a => String -> a -> a
+spy' msg a = trace (msg <> ": " <> toString (pShow a)) a
```
Separate PR

ch1bo replied:
https://github.com/cardano-scaling/hydra/pull/1504

ch1bo commented on spec/figures/incrementTx.pdf:
Should not be included here

ch1bo commented on spec/figures/fanoutTx.pdf:
Might be outdated

ch1bo commented on spec/figures/contestTx.pdf:
Might be outdated

ch1bo commented on testnets/cardano-configurations:
Separate PR?

ch1bo commented on spec/onchain.tex:
Rename not needed

v0d1ch committed:
Comment out failing coverage tests

We are not sure if we need to spend time on these right now so we
decided to leave those for now.

ch1bo commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -314,20 +314,19 @@ spec = around (showLogsOnFailure "DirectChainSpec") $ do
-                { snapshot
-                , signatures = aggregate [sign aliceSk snapshot]
-                }
+            postTx $ CloseTx headId headParameters (ConfirmedSnapshot{snapshot, signatures = aggregate [sign aliceSk snapshot]}) v toDecommit
```
Should: not require version in `CloseTx` again as it's already in `snapshot`

ch1bo commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -442,41 +442,55 @@ spec = around (showLogsOnFailure "DirectChainSpec") $ do
+                    , signatures = aggregate [sign aliceSk snapshot1]
+                    }
+                )
+                0
```
Should: not require `version` outside of `snapshot`

ch1bo commented on hydra-cluster/test/Test/ChainObserverSpec.hs:
```diff
@@ -51,13 +59,32 @@ spec = do
-                requestCommitTx hydraNode mempty >>= submitTx cardanoNode
+                commitTx <- requestCommitTx hydraNode commitUTxO
+
+                submitTx cardanoNode (signTx walletSk commitTx)
```
Unnecessary diff

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -192,6 +195,21 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
+        withClusterTempDir $ \tmpDir -> do
+          withCardanoNodeDevnet (contramap FromCardanoNode tracer) tmpDir $ \node ->
+            publishHydraScriptsAs node Faucet
+              >>= canFanoutWithDecommitRecorded tracer tmpDir node
```
Should: not require three integration tests (one is enough) and instead have these covered by `TxTrace` tests

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -566,6 +597,305 @@ initWithWrongKeys workDir tracer node@RunningNode{nodeSocket} hydraScriptsTxId =
+        case aliceChainConfig of
+          Direct DirectChainConfig{cardanoSigningKey} -> do
+            aliceSigningKey <- readFileTextEnvelopeThrow (AsSigningKey AsPaymentKey) cardanoSigningKey
+            let alicePVk = getVerificationKey aliceSigningKey
```
Should: use `keysFor`

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -566,6 +597,305 @@ initWithWrongKeys workDir tracer node@RunningNode{nodeSocket} hydraScriptsTxId =
+            [ TxOut walletAddress (lovelaceToValue decommitAmount) TxOutDatumNone ReferenceScriptNone
+            ]
+      -- here we set the change address to Alice to simplify the balance assertion in the end of test.
+      buildTransaction networkId nodeSocket aliceAddress commitUTxO1 (fst <$> UTxO.pairs commitUTxO1) decommitOutput >>= \case
```
Is this giving alice's hydra node the funds we decommit? .. or only the change?

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -566,6 +597,305 @@ initWithWrongKeys workDir tracer node@RunningNode{nodeSocket} hydraScriptsTxId =
+              expectSuccessOnSignedDecommitTx n1 headId walletSk body1 callDecommitHttpEndpoint
+
+              -- Decommit the second utxo
+              expectSuccessOnSignedDecommitTx n1 headId walletSk body2 callDecommitHttpEndpoint
```
Why decommit twice?

Failure cases and duplicate submissions are better covered by property tests (API or TxTrace tests?)

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -566,6 +597,305 @@ initWithWrongKeys workDir tracer node@RunningNode{nodeSocket} hydraScriptsTxId =
+  expectSuccessOnSignedDecommitTx n headId sk body httpCall = do
+    let signedDecommitTx = makeSignedTransaction [makeShelleyKeyWitness body (WitnessPaymentKey sk)] body
+    let signedDecommitClientInput = send n $ input "Decommit" ["decommitTx" .= signedDecommitTx]
+    join . generate $ oneof [pure signedDecommitClientInput, pure $ httpCall signedDecommitTx]
```
```suggestion
    join . generate $ elements [signedDecommitClientInput, httpCall signedDecommitTx]
```

and others too

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -225,6 +225,28 @@ aggregateCollectComObservation headId point blockNo currentHeads =
+aggregateDecrementObservation headId point blockNo currentHeads =
+  case findHeadState headId currentHeads of
+    Just headState ->
+      let newHeadState = headState{status = Open}
```
Follow up: total value locked decreased now and hence we need to (can) keep track of UTxO in the head here to fix the calculation

ch1bo replied:
Created issue: https://github.com/cardano-scaling/hydra-explorer/issues/3

ch1bo commented on hydra-node/bench/tx-cost/Main.hs:
```diff
@@ -228,6 +231,32 @@ costOfCollectCom = markdownCollectComCost <$> computeCollectComCost
           stats
 
+costOfDecommit :: IO Text
+costOfDecommit = markdownDecommitCost <$> computeDecommitCost
```
The transaction is called decrement

ch1bo commented on hydra-node/bench/tx-cost/TxCost.hs:
```diff
@@ -167,16 +184,19 @@ computeContestCost = do
     utxo <- arbitrary
-    (closedSnapshotNumber, _, stClosed@ClosedState{headId}) <- genStClosed ctx utxo
+    let (inHead, toDecommit) = splitUTxO utxo
+    (closedSnapshotNumber, inHead', toDecommit', stClosed@ClosedState{headId}) <- genStClosed ctx inHead (Just toDecommit)
```
Does not matter what is to be decommitted on contest, could keep it simple

ch1bo commented on hydra-node/bench/tx-cost/TxCost.hs:
```diff
@@ -167,16 +184,19 @@ computeContestCost = do
     let cp = ctxContestationPeriod ctx
     let contestUtxo = getKnownUTxO stClosed <> getKnownUTxO cctx
-    pure (unsafeContest cctx contestUtxo headId cp snapshot pointInTime, contestUtxo)
+    pure (unsafeContest cctx contestUtxo headId cp snapshot pointInTime version, contestUtxo)
```
Inconsistent version! Why possible?

ch1bo commented on hydra-node/exe/hydra-net/Main.hs:
```diff
@@ -224,7 +224,8 @@ injectReqSn peer snapshotNumber hydraKeyFile fakeHydraKeyFile = do
 
   client tracer sk party = Idle $ do
-    let msg = Data "2" (ReqSn @Tx snapshotNumber [])
+    let snapshotVersion = 0
```
Could: parameterize this on the command line

ch1bo commented on hydra-node/src/Hydra/Network/Message.hs:
```diff
@@ -77,6 +82,7 @@ data Message tx
     -- good idea to introduce the party in AckSn again or, maybe better, only
     -- the verification key of the party.
     AckSn {signed :: Signature (Snapshot tx), snapshotNumber :: SnapshotNumber}
+  | ReqDec {transaction :: tx, decommitRequester :: Party}
```
Should: not need decommitRequester

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -205,6 +235,10 @@ prepareServerOutput ServerOutputConfig{utxoInSnapshot} response =
+    DecommitRequested{} -> encodedResponse
+    DecommitApproved{} -> encodedResponse
+    DecommitFinalized{} -> encodedResponse
+    DecommitInvalid{} -> encodedResponse
```
Follow-up: this function should not be needed anymore (all txs are cbor)

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -278,10 +278,14 @@ test-suite tests
+    Hydra.Chain.Direct.Contract.Close.CloseCurrent
+    Hydra.Chain.Direct.Contract.Close.CloseInitial
+    Hydra.Chain.Direct.Contract.Close.CloseOutdated
+    Hydra.Chain.Direct.Contract.Close.Healthy
```
This looks like a lot of repetition

ch1bo commented on hydra-node/src/Hydra/API/HTTPServer.hs:
```diff
@@ -216,6 +223,15 @@ handleSubmitUserTx directChain body = do
+      pure $ responseLBS status400 [] (Aeson.encode $ Aeson.String $ pack err)
+    Right decommitTx -> do
+      putClientInput Decommit{decommitTx}
+      pure $ responseLBS status200 [] (Aeson.encode $ Aeson.String "OK")
```
Documented?

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -66,9 +67,27 @@ data PostChainTx tx
+      , headParameters :: HeadParameters
+      , confirmedSnapshot :: ConfirmedSnapshot tx
+      , version :: SnapshotVersion
+      }
```
Redundant `version` and `closeUTxOToDecommit`

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -449,8 +522,9 @@ closeTx ::
   OpenThreadOutput ->
   -- | Head identifier
   HeadId ->
+  SnapshotVersion ->
```
Add documentation why this is needed (or model it differently)

ch1bo commented on hydra-plutus/src/Hydra/Contract/Initial.hs:
```diff
@@ -1,8 +1,7 @@
 {-# LANGUAGE TemplateHaskell #-}
 {-# OPTIONS_GHC -fno-specialize #-}
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:conservative-optimisation #-}
```
Separate PR?

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -1,9 +1,8 @@
 {-# LANGUAGE DuplicateRecordFields #-}
 {-# LANGUAGE TemplateHaskell #-}
 {-# OPTIONS_GHC -fno-specialize #-}
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:conservative-optimisation #-}
```
Separate PR?

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -143,7 +144,8 @@ checkCollectCom ::
   mustCollectUtxoHash
-    && mustNotChangeParameters
+    && mustInitVersion
+    && mustNotChangeParameters (parties', parties) (contestationPeriod', contestationPeriod) (headId', headId)
```
Maybe not refactor? It's more explicit and requires the same amount of arguments as the body of the function is long

ch1bo replied:
Inlinable pragma actually makes this smaller than having it in where clauses.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -219,13 +209,61 @@ checkCollectCom ctx@ScriptContext{scriptContextTxInfo = txInfo} (contestationPer
 commitDatum :: TxOut -> [Commit]
 commitDatum input = do
-  let datum = getTxOutDatum input
+  let !datum = getTxOutDatum input
```
Why? plutus tx is already strict

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -249,17 +287,27 @@ checkClose ::
     && checkDeadline
-    && checkSnapshot
     && mustBeSignedByParticipant ctx headPolicyId
+    -- && checkLastKnownVersion
```
Must add this

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/FanOut.hs:
```diff
@@ -10,6 +10,7 @@ import Cardano.Api.UTxO as UTxO
 import Hydra.Chain.Direct.Contract.Mutation (Mutation (..), SomeMutation (..), changeMintedTokens)
 import Hydra.Chain.Direct.Fixture (slotLength, systemStart, testNetworkId, testPolicyId, testSeedInput)
 import Hydra.Chain.Direct.ScriptRegistry (genScriptRegistry, registryUTxO)
+import Hydra.Chain.Direct.State (splitUTxO)
```
Weird location for this

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -983,43 +1028,70 @@ genCollectComTx = do
+  let openUTxO = getKnownUTxO stOpen
+  pure (cctx, stOpen, unsafeDecrement cctx headId (ctxHeadParameters ctx) openUTxO snapshot signatures)
+
+splitUTxO :: UTxO -> (UTxO, UTxO)
```
What does it do? Explain how UTxO are split.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -614,19 +727,21 @@ fanoutTx ::
   -- | Snapshotted UTxO to fanout on layer 1
   UTxO ->
+  -- | Snapshotted decommit UTxO to fanout on layer 1
+  Maybe UTxO ->
```
Non-ideal type, can be empty in two variants (`Nothing` and `Just mempty`)

ch1bo commented on hydra-node/src/Hydra/HeadLogic/State.hs:
```diff
@@ -147,6 +147,11 @@ data CoordinatedHeadState tx = CoordinatedHeadState
   -- ^ Last seen snapshot and signatures accumulator. Spec: Û, ŝ and Σ̂
+  , decommitTx :: Maybe tx
+  -- ^ Client requested to decommit a 'UTxO' which is present in the
+  -- 'ConfirmedSnapshot'.
```
If its here, it's not present in a confirmed snapshot.

ch1bo commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -27,26 +27,43 @@ newtype SnapshotNumber
   -- ^ The set of transactions that lead to 'utxo'
+  , utxoToDecommit :: Maybe (UTxOType tx)
+  -- ^ UTxO to be decommitted. Spec: Ûω
+  -- TODO: what is the difference between Noting and (Just mempty) here?
```
Yes, should change to just `UTxOType tx` (which must be `Monoid`)

ch1bo commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -27,26 +27,43 @@ newtype SnapshotNumber
+  -- ^ UTxO to be decommitted. Spec: Ûω
+  -- TODO: what is the difference between Noting and (Just mempty) here?
+  -- | Snapshot version is 0 at start and is only bumped further on each
+  -- decommit that happens.
```
Find a better explanation from a snapshots perspective

ch1bo commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -55,32 +72,65 @@ instance IsTx tx => FromJSON (Snapshot tx) where
+--    utxo : {}
+--    utxoToDecommit : {}
+--    version : uint
+--   } / bytes
```
Is this a valid CDDL?

Also, this does not match the bytes produced below. So rather remove it than have it incorrect

ch1bo commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -55,32 +72,65 @@ instance IsTx tx => FromJSON (Snapshot tx) where
+        <> serialise (toData . toBuiltin $ toInteger number) -- CBOR(I(integer))
+        <> serialise (toData . toBuiltin $ hashUTxO @tx utxo) -- CBOR(B(bytestring)
+        <> serialise (toData . toBuiltin . hashUTxO @tx $ fromMaybe mempty utxoToDecommit) -- CBOR(B(bytestring)
+        <> serialise (toData . toBuiltin $ toInteger version) -- CBOR(I(integer))
```
Move version into first position?

ch1bo commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -126,8 +178,9 @@ instance IsTx tx => Arbitrary (ConfirmedSnapshot tx) where
+    utxoToDecommit <- arbitrary
     headId <- arbitrary
-    genConfirmedSnapshot headId 0 utxo ks
+    genConfirmedSnapshot headId 0 0 utxo utxoToDecommit ks
```
Order args consistently (version in front or back)

ch1bo commented on hydra-node/src/Hydra/HeadLogic/SnapshotOutcome.hs:
```diff
@@ -19,6 +19,8 @@ data NoSnapshotReason
   deriving stock (Eq, Show, Generic)
 
+-- REVIEW
+-- Spec: leader(s+1) = i
```
Pointless comment.

ch1bo commented on hydra-node/src/Hydra/HeadLogic/Outcome.hs:
```diff
@@ -69,6 +69,11 @@ data StateChanged tx
+  | DecommitRecorded
+      { decommitTx :: tx
+      , newLocalUTxO :: UTxOType tx
+      }
```
Very similar to `TransactionAppliedToLocalUTxO` .. could move up

ch1bo commented on hydra-node/src/Hydra/HeadLogic/Error.hs:
```diff
@@ -32,17 +32,21 @@ deriving anyclass instance (FromJSON (HeadState tx), FromJSON (Input tx), FromJS
   | SnapshotDoesNotApply {requestedSn :: SnapshotNumber, txid :: TxIdType tx, error :: ValidationError}
+  | DecommitTxInvalid {decommitTx :: tx, error :: ValidationError}
+  | DecommitTxInFlight {decommitTx :: tx}
+  | DecommitDoesNotApply {decommitTx :: tx, error :: ValidationError}
```
Only include `TxIdType`? Full tx should be available to the party, so simple txid is enough.

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -693,14 +791,16 @@ createMockNetwork node nodes =
 -- NOTE: This implementation announces hard-coded contestationDeadlines. Also,
 -- all heads will have the same 'headId' and 'headSeed'.
-toOnChainTx :: UTCTime -> PostChainTx tx -> OnChainTx tx
+toOnChainTx :: Monoid (UTxOType tx) => UTCTime -> PostChainTx tx -> OnChainTx tx
```
Redundant constraint?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -275,15 +258,47 @@ genCollectComMutation (tx, _utxo) =
+    pure $ valueFromList [(assetId, q)]
+  -- Add another output which would extract the 'removedValue'. The ledger
+  -- would check for this, and this is needed because the way we implement
+  -- collectCom checks.
```
This is not only used in collectcom now, should generalize the comments / applicability of this too.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -275,15 +258,47 @@ genCollectComMutation (tx, _utxo) =
+        Head.Open{Head.parties = parties, snapshotNumber, contestationPeriod, headId, version, utxoHash}
       st -> error $ "Unexpected state " <> show st
 
   mutateUTxOHash = do
```
Used only once, inline

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -275,15 +258,47 @@ genCollectComMutation (tx, _utxo) =
+      Head.Open{Head.utxoHash = toBuiltin mutatedUTxOHash, parties, contestationPeriod, headId, snapshotNumber, version}
     st -> st
+
+  mutateVersion = do
```
Used only once, inline

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -87,6 +87,7 @@ type UTxOWithScript = (TxIn, TxOut CtxUTxO, HashableScriptData)
 
 newtype UTxOHash = UTxOHash ByteString
   deriving stock (Eq, Show, Generic)
+  deriving newtype (Semigroup, Monoid)
```
Unexpected, where do we `mappend` UTxOHashes? this is likely not a good idea

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -363,27 +369,33 @@ prepareTxToPost timeHandle wallet ctx spendableUTxO tx =
         Right collectTx -> pure collectTx
-    CloseTx{headId, headParameters, confirmedSnapshot} -> do
+    DecrementTx{headId, headParameters, snapshot, signatures} ->
+      case decrement ctx headId headParameters spendableUTxO snapshot signatures of
```
Why no `ConfirmedSnapshot`?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -165,6 +168,7 @@ data ChainTransition
   | Abort
   | Commit
   | Collect
+  | Decrement
```
Are these still used?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -464,6 +468,42 @@ collect ctx headId headParameters utxoToCollect spendableUTxO = do
+  | SnapshotDecrementUTxOIsNull
+  deriving stock (Show)
+
+decrement ::
```
Haddocks

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -464,6 +468,42 @@ collect ctx headId headParameters utxoToCollect spendableUTxO = do
+  HeadId ->
+  HeadParameters ->
+  -- | Spendable UTxO containing head, initial and commit outputs
+  UTxO ->
```
Re-order args (generic first)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -413,6 +416,74 @@ collectComTx networkId scriptRegistry vk headId headParameters (headInput, initi
+    emptyTxBody
+      & addInputs [(headInput, headWitness)]
+      & addReferenceInputs [headScriptRef]
+      -- NOTE: at this point 'utxoToDecommit' is populated
```
What does this mean?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -413,6 +416,74 @@ collectComTx networkId scriptRegistry vk headId headParameters (headInput, initi
+        { signature = toPlutusSignatures signatures
+        , numberOfDecommitOutputs =
+            fromIntegral $ length $ maybe [] toList utxoToDecommit
+        }
```
Should contain snapshot number (as not in head datum)

ch1bo replied:
Also update spec todo (which claims this diff)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -363,27 +369,33 @@ prepareTxToPost timeHandle wallet ctx spendableUTxO tx =
+        Right decrementTx' -> pure decrementTx'
+    CloseTx{headId, headParameters, confirmedSnapshot, version, closeUTxOToDecommit} -> do
+      unless (fromMaybe mempty (Snapshot.utxoToDecommit $ getSnapshot confirmedSnapshot) == closeUTxOToDecommit) $
+        throwIO (FailedToConstructCloseTx @Tx)
```
Why `closeUTxOToDecommit` separate from `confirmedSnapshot`?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -490,17 +532,19 @@ close ctx spendableUTxO headId HeadParameters{parties, contestationPeriod} confi
           , openParties = partyToChain <$> parties
           }
-  pure $ closeTx scriptRegistry ownVerificationKey closingSnapshot startSlotNo pointInTime openThreadOutput headId
+  pure $ closeTx scriptRegistry ownVerificationKey closingSnapshot startSlotNo pointInTime openThreadOutput headId offChainVersion
```
Should: update `CloseWithConfirmedSnapshot` to reflect the three cases from the spec

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -478,10 +552,25 @@ closeTx scriptRegistry vk closing startSlotNo (endSlotNo, utcTime) openThreadOut
+
   headRedeemer =
     toScriptData
       Head.Close
```
Sum type in close redeemer

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -249,17 +287,27 @@ checkClose ::
  where
+  checkSnapshotNumber =
+    traceIfFalse $(errorCode TooOldSnapshot) $
+      closedSnapshotNumber >= snapshotNumber
```
Not needed

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -356,17 +407,21 @@ checkContest ::
-checkContest ctx contestationDeadline contestationPeriod parties closedSnapshotNumber sig contesters headId =
+checkContest ctx contestationDeadline contestationPeriod parties closedSnapshotNumber sig contesters headId inputVersion expectedVersion utxoToDecommitHash =
   mustNotMintOrBurn txInfo
+    -- && checkLastKnownVersion
```
Must add this

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -272,27 +320,36 @@ checkClose ctx parties initialUtxoHash sig cperiod headPolicyId =
+        -- would be invalid anyway?
+        (utxoToDecommitHash, outputVersion)
 
   checkSnapshot
```
Convert to case separation as in spec

ch1bo commented on hydra-plutus/src/Hydra/Contract/Commit.hs:
```diff
@@ -1,8 +1,7 @@
 {-# LANGUAGE TemplateHaskell #-}
 {-# OPTIONS_GHC -fno-specialize #-}
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:conservative-optimisation #-}
```
Separate PR?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Decrement.hs:
```diff
@@ -0,0 +1,211 @@
+    -- different (not preserved) from the head.
+    --
+    -- Ensures values are preserved between head input and output.
+    ChangeValueInOutput
```
This and the mutation above do the same thing

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Decrement.hs:
```diff
@@ -0,0 +1,211 @@
+        pure $ ChangeOutput 0 $ modifyInlineDatum (replaceParties mutatedParties) headTxOut
+    , -- New version v′ is incremented correctly
+      SomeMutation (pure $ toErrorCode IncorrectVersion) UseDifferentSnapshotVersion <$> do
+        mutatedSnapshotVersion <- arbitrarySizedNatural `suchThat` (> healthySnapshotVersion + 1)
```
Should invalidate as long as it's not `== healthySnapshotVersion + 1`

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close/Healthy.hs:
```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
```
Not needed

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/ContractSpec.hs:
```diff
@@ -113,17 +116,27 @@ spec = parallel $ do
+      propTransactionEvaluates healthyCloseOutdatedTx
+    prop "does not survive random adversarial mutations" $
+      propMutation healthyCloseOutdatedTx genCloseOutdatedMutation
+  describe "ContestCurrent" $ do
```
Combinatorial / maintenance explosion right here.. maybe have a (labelled) healthy case generator and define mutations more introspective?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/ContractSpec.hs:
```diff
@@ -203,27 +216,33 @@ prop_hashingCaresAboutOrderingOfTxOuts =
+          snapshotVersion = toInteger version
+          utxoHash = toBuiltin (hashUTxO @SimpleTx utxo)
+          utxoToDecommitHash = (toBuiltin . hashUTxO @SimpleTx $ fromMaybe mempty utxoToDecommit)
+       in verifySnapshotSignature onChainParties (headIdToCurrencySymbol headId) snapshotNumber utxoHash utxoToDecommitHash snapshotVersion signatures
```
Argument ordering not intuitive

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close/CloseCurrent.hs:
```diff
@@ -0,0 +1,274 @@
+    , -- , -- XXX: Last known open state version is recorded in closed state
+      --   SomeMutation (pure $ toErrorCode LastKnownVersionIsNotMatching) MutateSnapshotVersion <$> do
+      --     mutatedSnapshotVersion <- arbitrarySizedNatural `suchThat` (> healthyCloseSnapshotVersion)
+      --     pure $ ChangeOutput 0 $ modifyInlineDatum (replaceSnapshotVersionInClosed $ toInteger mutatedSnapshotVersion) headTxOut
```
Commented mutation!?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -517,14 +561,15 @@ contest ::
       ?> CannotFindHeadOutputToContest
   closedThreadOutput <- checkHeadDatum headUTxO
-  pure $ contestTx scriptRegistry ownVerificationKey sn sigs pointInTime closedThreadOutput headId contestationPeriod
+  pure $ contestTx scriptRegistry ownVerificationKey sn sigs pointInTime closedThreadOutput headId contestationPeriod offChainVersion
```
Pull `ContestType` case separation outside (into head logic even?)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -958,18 +1011,10 @@ genCommits' genUTxO ctx txInit = do
 genCommitFor :: VerificationKey PaymentKey -> Gen UTxO
-genCommitFor vkey =
-  frequency
-    [ (1, pure mempty)
```
Why no empty commits anymore? This seems to be used in `forAllCommit'` in `StateSpec`

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -983,43 +1028,70 @@ genCollectComTx = do
-  snapshot <- genConfirmedSnapshot headId 0 u0 (ctxHydraSigningKeys ctx)
+  let (confirmedUtxo, utxoToDecommit) = splitUTxO u0
+  let version = 0
+  snapshot <- genConfirmedSnapshot headId 1 version confirmedUtxo (Just utxoToDecommit) (ctxHydraSigningKeys ctx)
```
Rather update generator to produce snapshots that sometimes decommit something (from a given total utxo)

ch1bo replied:
This is hard, because the generator is defined generically over `tx`. Maybe we should have a generic one to use in `arbitrary` and a specific one for these kinds of tests? Anyways.. not now.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -983,43 +1028,70 @@ genCollectComTx = do
   (u0, stOpen@OpenState{headId}) <- genStOpen ctx
-  confirmed <- genConfirmedSnapshot headId 0 u0 []
+  let (confirmedUtXO, utxoToDecommit) = splitUTxO u0
+  let version = 1
```
Why version `1`?

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -73,12 +72,14 @@ headValidator oldState input ctx =
+      checkDecrement ctx parties snapshotNumber contestationPeriod headId version signature numberOfDecommitOutputs
+    (Open{parties, utxoHash = initialUtxoHash, contestationPeriod, headId, snapshotNumber, version}, Close{signature, version = expectedVersion, utxoToDecommitHash}) ->
+      checkClose ctx parties initialUtxoHash signature contestationPeriod headId snapshotNumber version expectedVersion utxoToDecommitHash
+    (Closed{parties, snapshotNumber = closedSnapshotNumber, contestationDeadline, contestationPeriod, headId, contesters, version}, Contest{signature, version = expectedVersion, utxoToDecommitHash}) ->
```
Must not take `expectedVersion` from redeemer, but check in datums!

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -73,12 +72,14 @@ headValidator oldState input ctx =
+    (Closed{parties, snapshotNumber = closedSnapshotNumber, contestationDeadline, contestationPeriod, headId, contesters, version}, Contest{signature, version = expectedVersion, utxoToDecommitHash}) ->
+      checkContest ctx contestationDeadline contestationPeriod parties closedSnapshotNumber signature contesters headId version expectedVersion utxoToDecommitHash
+    (Closed{parties, utxoHash, utxoToDecommitHash, contestationDeadline, headId}, Fanout{numberOfFanoutOutputs, numberOfDecommitOutputs}) ->
+      checkFanout utxoHash utxoToDecommitHash contestationDeadline numberOfFanoutOutputs numberOfDecommitOutputs ctx headId parties
```
Introduce redeemer sub-types to avoid errors!

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -356,17 +407,21 @@ checkContest ::
     && mustPushDeadline
-    && mustNotChangeParameters
+    && mustNotChangeParameters (parties', parties) (contestationPeriod', contestationPeriod) (headId', headId)
     && mustPreserveValue
```
Should order constraints like in spec

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -413,6 +416,74 @@ collectComTx networkId scriptRegistry vk headId headParameters (headInput, initi
+      Head.Decrement
+        { signature = toPlutusSignatures signatures
+        , numberOfDecommitOutputs =
+            fromIntegral $ length $ maybe [] toList utxoToDecommit
```
Check that we don't agree to decrement 0 utxos off-chain (to avoid spam)

ch1bo replied:
Created follow-up issue: https://github.com/cardano-scaling/hydra/issues/1502

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/FanOut.hs:
```diff
@@ -105,27 +112,64 @@ data FanoutMutation
+      SomeMutation (pure $ toErrorCode BurntTokenNumberMismatch) MutateThreadTokenQuantity <$> do
+        (token, _) <- elements burntTokens
+        changeMintedTokens tx (valueFromList [(token, 1)])
+
```
Unnecessary diff

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/FanOut.hs:
```diff
@@ -105,27 +112,64 @@ data FanoutMutation
+        let noOfUtxoToOutputs = fromIntegral . size $ toMap (fst healthyFanoutSnapshotUTxO)
+        let noOfUtxoDecommitToOutputs = fromIntegral . size $ toMap (snd healthyFanoutSnapshotUTxO)
+        n <- elements [1 .. 3]
+        pure (Head.Fanout (noOfUtxoToOutputs - n) noOfUtxoDecommitToOutputs)
```
Maybe just pick the simplest mutation:
- add output
- modify output
- drop output
- change redeemer (number of considered outputs)

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -426,6 +425,14 @@ prop_observeAnyTx =
+      getLovelace a =
+        sum $ selectLovelace . txOutValue . snd <$> UTxO.pairs a
+   in getLovelace utxo === getLovelace (inHead <> toDecommit)
+        .&&. UTxO.difference (inHead <> toDecommit) utxo === mempty
```
Test re-implements the function

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -773,8 +948,18 @@ hasNoEffectSatisfying outcome predicate =
   [TxIdType tx] ->
   Snapshot tx
-testSnapshot = Snapshot testHeadId
+testSnapshot number version utxo confirmed =
```
Always used with version = 0

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -502,16 +667,20 @@ spec =
+        st <-
+          run $
+            runHeadLogic bobEnv ledger st0 $ do
+              step (receiveMessage $ ReqSn 0 1 [] Nothing)
```
Argument order is important! version is first here, make it consistent like this everywhere

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -395,16 +443,12 @@ genInit hydraParties = do
-genPayment :: WorldState -> Gen (Party, Payment)
-genPayment WorldState{hydraParties, hydraState} =
+genPayment :: CardanoSigningKey -> WorldState -> Gen (Party, Payment)
+genPayment to WorldState{hydraParties, hydraState} =
```
Generalization not needed (target is also picked arbitrarily in genDecommit)

ch1bo commented on hydra-node/test/Hydra/NodeSpec.hs:
```diff
@@ -86,16 +86,18 @@ spec = parallel $ do
-                `shouldThrow` \(_ :: ParameterMismatch) -> True
+          env
+            /= testEnvironment
+            ==> do
```
unnecessary diff

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -373,6 +373,93 @@ spec = parallel $ do
+                waitUntilMatch [n1] $
+                  \case
+                    GetUTxOResponse{headId, utxo} -> headId == testHeadId && not (member 42 utxo)
+                    _ -> False
```
Move negative tests (rejection) from end-to-end test here?

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -945,6 +1299,8 @@ aggregate st = \case
                   , localTxs = mempty
                   , confirmedSnapshot = InitialSnapshot{headId, initialUTxO}
```
Spec: `S̄ ← snObj(0, 0, U0 , ∅, ⊥)` .. maybe this is an indicator that having a single product type is better and unifies initial/confirmed cases?

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+            wait $ WaitOnNotApplicableDecommitTx decommitTx
+        | otherwise ->
+            -- REVIW: cause . ClientEffect $ ServerOutput.DecommitInvalid
+            Error $ RequireFailed $ DecommitTxInFlight{decommitTx = existingDecommitTx}
```
Should be a decommit invalid like (analogous to ReqTx)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+      Nothing -> continue
+
+  checkValidDecommitTx cont =
+    case applyTransactions ledger currentSlot confirmedUTxO [decommitTx] of
```
Why check against confirmed ledger? The `ReqDec` needs to apply to the seen ledger later anyways.. so why not also check it against that here?

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+--
+-- This way we are checking _later_ than we could but it will allow us to not
+-- discard decommit tx which is maybe part of the snapshot _in flight_ but we
+-- just didn't see the `AckSn` for it yet.
```
Exactly, so why not check against the local view, which will be the desired, to be confirmed, state if we are the snapshot leader.

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -106,23 +117,156 @@ spec =
       it "confirms snapshot given it receives AckSn from all parties" $ do
-        let reqSn = receiveMessage $ ReqSn 1 []
-            snapshot1 = Snapshot testHeadId 1 mempty []
+        -- TODO: perhaps use smart constructor for ReqSn to reduce the noise
```
```suggestion
```

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -106,23 +117,156 @@ spec =
+              input = receiveMessage reqDec
+              st = inOpenState threeParties
+              outcome = update aliceEnv ledger st input
+           in outcome
```
Only tests that uses `let`/`in` (other `do`/`let`)

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -106,23 +117,156 @@ spec =
+                  ClientEffect DecommitRequested{headId, utxoToDecommit} ->
+                    headId == testHeadId && utxoToDecommit == utxoRef 1
+                  NetworkEffect ReqDec{transaction, decommitRequester} ->
+                    transaction == decommitTx && decommitRequester == alice
```
This is not expected!

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -106,23 +117,156 @@ spec =
+            step reqDecEvent1
+            getState
+
+          let outcome = update bobEnv ledger s1 reqDecEvent2
```
Why switch environment? Also simplify

```suggestion
          outcome <- runHeadLogic bobEnv ledger s0 $ do
            step reqDecEvent1
            step reqDecEvent2
```

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -106,23 +117,156 @@ spec =
+            _ -> False
+
+        -- REVIEW: check if duplicated with -> "wait for second decommit"
+        it "waits if a requested decommit tx is not (yet) applicable due to already in flight" $ do
```
Yes, this is the same. But expressed in a more whitebox-style (preparing the open state specifically)

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -106,23 +117,156 @@ spec =
+
+          s2 `shouldSatisfy` \case
+            (Open OpenState{coordinatedHeadState = CoordinatedHeadState{decommitTx}}) -> decommitTx == Just decommitTx'
+            _ -> False
```
Why test so low-level?

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -106,23 +117,156 @@ spec =
+
+          s0 `shouldSatisfy` \case
+            (Open OpenState{coordinatedHeadState = CoordinatedHeadState{decommitTx}}) -> isNothing decommitTx
+            _ -> False
```
Why assert the starting state (which is specified one line above!?)

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -106,23 +117,156 @@ spec =
+          let reqDecEvent = receiveMessage ReqDec{transaction = decommitTx', decommitRequester = alice}
+
+          update aliceEnv ledger s0 reqDecEvent
+            `assertWait` WaitOnNotApplicableDecommitTx decommitTx'
```
Move up, next to the other wait not applicable of reqdec

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+    , currentSlot
+    } = openState
+
+-- ** Decrementing funds from the Head
```
Check haddock outline (function order)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+onOpenChainDecrementTx :: IsTx tx => Environment -> OpenState tx -> Outcome tx
+onOpenChainDecrementTx Environment{party} openState
+  | -- Spec: if outputs(txω) = 𝑈ω
+    -- REVIEW: should we get Uω from observation instead of local state?
```
Yes, this is wrong. The equality below is always true (unless we fucked up our state)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -308,21 +318,17 @@ onOpenNetworkReqTx ::
-  -- Spec: Tall ← ̂Tall ∪ { (hash(tx), tx) }
+  -- REVIEW: would it brake the retry mechanism if we produce this inside `waitApplyTx`?
+  -- Spec Gap: this event is missing
   (newState TransactionReceived{tx} <>) $
```
Resolve comment. This is missing because the spec does not do the transaction hashing (optimization) anymore.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+            Error $ RequireFailed $ DecommitTxInFlight{decommitTx = existingDecommitTx}
+
+  maybeEmitSnapshot =
+    if isLeader parameters party nextSn
```
Only do this if no snapshot in flight!?

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -384,43 +401,80 @@ onOpenNetworkReqSn ::
+    | otherwise =
+        continue
+
+  requireReqSnbr continue
```
Combine requirements
```suggestion
    | sv /= version =
        Error $ RequireFailed $ ReqSvNumberInvalid{requestedSv = sv, lastSeenSv = version}
```

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -439,13 +493,28 @@ onOpenNetworkReqSn env ledger st otherParty sn requestedTxIds =
+            -- Spec: ηω ← combine(outputs(txω))
+            let utxoToDecommit = utxoFromTx decommitTx
+            -- Spec: 𝑈_active ← S⁻.𝑈 ◦ txω
+            let activeUTxO = newConfirmedUTxO `withoutUTxO` utxoToDecommit
```
Fix spec: the `\ outputs(tx_omega)` is missing

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -384,43 +401,80 @@ onOpenNetworkReqSn ::
+        waitResolvableTxs $ do
+          -- Spec Gap: this is missing
+          -- Spec: Treq ← {T̂ [h] | h ∈ Treq#}
+          let requestedTxs = mapMaybe (`Map.lookup` allTxs) requestedTxIds
```
Remove spec gap (it's the txid optimization) and maybe combine into `waitResolveTxs $ \requestedTxs ->`

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -384,43 +401,80 @@ onOpenNetworkReqSn ::
+          let requestedTxs = mapMaybe (`Map.lookup` allTxs) requestedTxIds
+          -- Spec: require 𝑈_active ◦ Treq ≠ ⊥
+          --       𝑈 ← 𝑈_active ◦ Treq
+          --       η ← combine(𝑈)
```
This is implicit in `sign nextSnapshot`

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -384,43 +401,80 @@ onOpenNetworkReqSn ::
+                -- Spec Gap: missing ∀tx ∈ Treq : output (seen, tx)
+                --                   L̂ ← 𝑈
+                --                   𝑋 ← T̂
+                --                   T̂ ← ∅
```
No. This is the `pruneTransactions` algorithm. Also, `output(seen, tx)` is basically `newState SnapshotRequested { ... requestedTxIds}`

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -506,15 +575,30 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
-      -- Spec: (j,.) ∉ ̂Σ
+      -- Spec: require (j,⋅) ∉ Σ̂
       requireNotSignedYet sigs $ do
+        -- Spec Gap: missing Σ̂[j] ← σⱼ
```
No, this is in `ifAllMembersHaveSigned` (it produces a new list of signatures `sigs'`)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -506,15 +575,30 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+              -- Spec: if txω ≠ ⊥
+              --         postTx (decrement, vˆ, ŝ, η, ηω)
+              --         output (conf, txω )
+              & maybeEmitDecrementTx snapshot multisig
```
```suggestion
              & maybePostDecrementTx snapshot multisig
```

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -506,15 +575,30 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+          -- Spec Gap: missing
+          --         η ← combine(𝑈ˆ)
+          --         ηω ← combine(outputs(txω ))
+          -- Spec: require MS-Verify(k ̃H, (cid‖v^‖ŝ‖η‖ηω), σ̃)
```
Nothing missing. It's all included in `requireVerifiedMultisignature` of `snapshot`

This indicates that we should "factor out" signature creation and verification for a particular snapshot object in the spec pseudocode. i.e. this would mean that we have a seen $\hat{\mathcal S} \gets \mathsf{snObj}(...)$ in the spec.

ch1bo replied:
There is a difference here: In the spec we verify using $tx_\omega$ while in the code we use the `seenSnapshot` (resolved in `waitOnSeenSnapshot` into `snapshot`). Don't think this is a problem, but it could be cleare if we indeed do $\hat{\mathcal S} \gets \mathsf{snObj}(...)$ in the spec.

ch1bo replied:
Not done right now, maybe later .. within #199

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
         outcome
           <> newState SnapshotRequestDecided{snapshotNumber = nextSn}
-          <> cause (NetworkEffect $ ReqSn nextSn (txId <$> localTxs))
+          -- Spec Gap: how S⁻.s + 1 is calculated is not the same as nextSn
```
Could make it more similar, but line 576 `wait ŝ = s` and 591 (new) `S⁻ ← snObj(vˆ, ŝ, 𝑈ˆ, T̂ , txω)` implies that it's the same

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -586,12 +883,27 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+            parameters
+            confirmedSnapshot
+            version
+            (fromMaybe mempty $ Hydra.Snapshot.utxoToDecommit $ getSnapshot confirmedSnapshot)
```
This goes full circle now. Make `utxoToDecommit` a `UTxOType` directly

ch1bo replied:
Hm.. this has pros/cons both ways.

ch1bo replied:
Decide to keep it as is (for now).

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -586,12 +883,27 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+  --       η ← combine(S⁻.𝑈)
+  --       ηω ← combine(outputs(S⁻.txω))
+  --       ξ ← S⁻.σ
+  --       postTx (close, S⁻.v, S⁻.s, η, ηω,ξ)
```
We basically pass the `snObj` to close tx construction. However, we should update the code and spec to pick the right `CloseType` (or `CloseAction` case) here.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -610,15 +923,22 @@ onOpenChainCloseTx ::
+    <> cause notifyClient
+    <> ( -- Spec: if S⁻.s > sc
+         --          postTx (contest, S⁻.v, S⁻.s, η, ηω, ξ)
+         if doContest
```
maybeContest?

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -610,15 +923,22 @@ onOpenChainCloseTx ::
+  -- Spec Gap: out of order & missing?
+  --       η ← combine(S⁻.𝑈)
+  --       ηω ← combine(outputs(S⁻.txω))
+  --       ξ ← S⁻.σ
```
Nothing missing as long as we give the snapshot to contestTx

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -648,11 +969,17 @@ onClosedChainContestTx ::
     <> if
-      | snapshotNumber < number (getSnapshot confirmedSnapshot) ->
+      | -- Spec: if S⁻.s > sc
+        snapshotNumber < number (getSnapshot confirmedSnapshot) ->
```
Align with spec

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -987,7 +1343,37 @@ aggregate st = \case
           sigs = Map.insert party signature signatories
       _otherState -> st
-  HeadIsReadyToFanout{} ->
+  DecommitRecorded decommitTx newLocalUTxO ->
```
Record syntax

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+    (utxoFromTx <$> decommitTx) == utxoToDecommit =
+      -- Spec: txω ← ⊥
+      --       vˆ  ← v
+      newState DecommitFinalized
```
Observe new version from chain (to only have + 1 in a single place -> when constructing the decrement tx)

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -200,10 +198,41 @@ instance StateModel WorldState where
+                ]
+                  <> ([(2, genDecommit) | length confirmedUTxO > 1])
+               | not (null confirmedUTxO)
+               ]
```
`if`, `then`, `else` is easier here

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -223,20 +252,24 @@ instance StateModel WorldState where
+  precondition WorldState{hydraState = Closed{headParameters = HeadParameters{parties}}} (Fanout p) =
+    p `elem` parties
+  precondition WorldState{hydraState = Open{headParameters = HeadParameters{parties}}} (CloseWithInitialSnapshot p) =
+    p `elem` parties
```
Why are these ``party `elem` parties`` needed now?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -154,6 +155,7 @@ instance StateModel WorldState where
     -- different return values.
     Init :: Party -> Action WorldState ()
     Commit :: Party -> UTxOType Payment -> Action WorldState ActualCommitted
+    Decommit :: Party -> Payment -> Action WorldState ()
```
Modelling a decommit to also just select a `UTxOType Payment` to remove could be simpler?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -531,13 +577,19 @@ instance
+        case hydraState of
+          Open{offChainState = OffChainState{confirmedUTxO}} -> do
+            let tx = lookup var
+            when ((to tx, value tx) `List.elem` confirmedUTxO) $ do
```
Why did this change?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxTraceSpec.hs:
```diff
@@ -58,52 +87,173 @@ prop_traces =
+        & cover 0.5 (fanoutWithEmptyUTxO steps) "fanout with empty UTxO"
+        & cover 5 (fanoutWithSomeUTxO steps) "fanout with some UTxO"
+        -- & cover 0.5 (fanoutWithDecrement steps) "fanout with something to decrement"
+        -- & cover 0.5 (fanoutWithSomeUTxOAndDecrement steps) "fanout with some UTxO and something to decrement"
```
Commented code

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxTraceSpec.hs:
```diff
@@ -113,80 +263,253 @@ data State
+                  actor <- elements allActors
+                  snapshot <- genSnapshot
+                  pure $ Some Contest{actor, snapshot}
+               ]
```
```suggestion
          , do
              actor <- elements allActors
              snapshot <- genSnapshot
              pure $ Some Contest{actor, snapshot}
          ]
```

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxTraceSpec.hs:
```diff
@@ -113,80 +263,253 @@ data State
+        , do
+            -- more in head
+            utxoInHead' <- increaseValues utxoInHead
+            pure validSnapshot{snapshotUTxO = utxoInHead'}
```
Aren't these invalid? Do we really want to create invalid transitions? IIRC this test suite specializes on testing sequences of otherwise valid steps

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxTraceSpec.hs:
```diff
@@ -113,80 +263,253 @@ data State
+    Fanout{snapshot} ->
+      headState == Closed
+        && snapshotUTxO snapshot == utxoInHead
+        && decommitUTxO snapshot == decommitUTxOInHead
```
This is the same as the `precondition`?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxTraceSpec.hs:
```diff
@@ -113,80 +263,253 @@ data State
+
+balanceUTxOInHead :: Ord k => Map k Natural -> Map k Natural -> Map k Natural
+balanceUTxOInHead currentUtxoInHead someUTxOToDecrement =
+  currentUtxoInHead `Map.difference` someUTxOToDecrement
```
Using `\\` is even shorter than to spell out `balanceUTxOInHead`

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxTraceSpec.hs:
```diff
@@ -197,76 +520,192 @@ instance HasVariables (Action Model a) where
+
+-- | Application monad to perform model actions. Currently it only keeps a
+-- 'UTxO' which is updated whenever transactions are valid in 'performTx'.
+newtype AppM a = AppM {runAppM :: ReaderT (IORef (UTxO, Natural)) IO a}
```
Why is a Natural (version) kept in the monad state?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxTraceSpec.hs:
```diff
@@ -197,76 +520,192 @@ instance HasVariables (Action Model a) where
+    case action of
+      Decrement{actor, snapshot} -> do
+        (_, v) <- get
+        tx <- newDecrementTx actor (signedSnapshot snapshot (UnsafeSnapshotVersion v))
```
Not sure if keeping the version in the run model is a good idea. I think we are even interested in modelling how various traces of snapshot versions are working (e.g. can't decrement twice with the same version; can close with current - 1 or current version, but no other version; ...)

ch1bo reviewed (commented):
Leaving *a few* comments for myself

ch1bo committed:
Not update cardano-configurations submodule

ch1bo committed:
Drop unused incrementTx figure from spec

ch1bo committed:
Revert fanout m' renaming

ch1bo committed:
Rename tx cost benchmark decommit -> decrement

ch1bo committed:
Don't use splitUTxO and use version 0 in tx-cost benchmark

ch1bo committed:
Move sequence diagram into dedicated docs page

ch1bo committed:
Must not change version in close/contest

Enabled mutation tests and updated scripts

ch1bo committed:
Limit size of generated contesters to reasonable values

ch1bo committed:
Broaden version mutation for collect

ch1bo committed:
Broaden version mutation for decrement

ch1bo committed:
Introduce CloseRedeemer and ContestRedeemer

This further aligns implementation with the spec and makes for a clearer
case separation.

ch1bo committed:
Use redeemers in tx construction

ch1bo committed:
Use redeemers in mutation tests

ch1bo committed:
Remove a now invalid mutation test

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+  partyIsLeader = isLeader parameters party nextSn && not (null localTxs)
+
+  maybeEmitSnapshot outcome =
+    if seenSn == confirmedSn && partyIsLeader
```
Fix also in spec: only snapshot if there is something to snapshot (i.e. localTxs is not null)

ch1bo replied:
@v0d1ch @ffakenz this is what you raised on slack. I realize that this is fine in the spec as-is: The snapshot leader would request a snapshot with the `decommitTx` being cleared (and any `localTxs` if there are any). Why? This makes the next decommit possible.

ch1bo replied:
I've done this now here: https://github.com/input-output-hk/hydra/commit/1f166b78e5e16c2be1758151127bfc2aac45204a

ch1bo committed:
Fix close transaction to use last open state version in output

The healthy case of close "outdated" mutations was also not correctly
setting up the versions.

ch1bo committed:
Add newVersion to OnDecrementTx and fix off-chain protocol

Seems like the previous practice of incrementing version in HeadLogic
was prone to race conditions.

ch1bo committed:
Observe distributed outputs from decrementTx

This allows us to align the HeadLogic part about clearing pending
decommits more faithfully.

Required to add a new method `outputsOfUTxO` to `IsTx` which feels to
become a bit "too big" for a type class.

ch1bo committed:
Remove snapshot number from open state and introduce sub-types

ch1bo committed:
Do not expect decrements with snapshot number 0 to fail

ch1bo committed:
Use sub-types for decrement/close in head validator

ch1bo committed:
Remove redundant strictness annotations in Head script

ch1bo committed:
Rename and re-number head errors

ch1bo committed:
Emit snapshot with cleared decommitTx onDecrementTx

ch1bo committed:
Sub-type closedDatum and make deltaHash a Maybe

ch1bo committed:
Re-order Snapshot fields and align signable representation with spec

ch1bo committed:
Add TODO about preventing empty decommits

ch1bo committed:
Change Decommit server outputs to be more in-line with TxValid/TxInvalid

This resolves a TODO (REVIEW) comment

ch1bo committed:
Regenerate golden files and make XXX comments mention Spec instead

ch1bo committed:
Fix contest off-chain

ch1bo committed:
Document and re-order arguments of 'decrement'

ch1bo committed:
Remove unused Semigroup and Monoid of UTxOHash

ch1bo committed:
Simplify decommit how-to

ch1bo committed:
Expand changelog entry

ch1bo committed:
Remove decommitRequester

ch1bo committed:
Update haddocks and drop a redundant test

ch1bo committed:
Check decommit request against local ledger

The explanation of why checking against the confirmed ledger on client
submission does not hold up. We can (and should) check against the local
ledger state in both cases as it's our local node's view of what is
valid and what it would snapshot next anyways.

ch1bo committed:
Only request snapshot on ReqDec if no snapshot in flight

ch1bo committed:
Resolve some spec gap comments in head logic

ch1bo committed:
Make AckSn processing emit a PartySignedSnapshot

This matches with the Σ̂[j] ← σj assignment from the pseudo-code.
This does not change the semantics, as the SnapshotConfirmed state
changed event includes the full multi-signature anyways.

ch1bo committed:
Align HeadLogic closer with spec

ch1bo committed:
Combine requires when processing a ReqSn

ch1bo committed:
Add DecommitRecorded and DecommitFinalized to genStateChanged

ch1bo committed:
Remove unused SnapshotOutcome data types (and module)

ch1bo committed:
Consolidate Wait outcomes for decommit request processing

ch1bo committed:
Not assert starting state in HeadLogicSpec

ch1bo committed:
Remove redundant language pragmas

ch1bo committed:
Refactor CloseTx/close/closeTx and add docs to openVersion

ch1bo committed:
Also refactore ContestTx/contest/contestTx to match close

ch1bo committed:
Also refactor DecrementTx/decrement/decrementTx to take a ConfirmedSnapshot

This provides a more uniform interface across the chain layer.

ch1bo committed:
Untangle some healthyXXX functions

ch1bo committed:
Review and unmark red on-chain spec areas

This includes a re-ordering of clauses in checkDecrement, which did not
change semantics (but script hashes).

ch1bo committed:
Clear gaps and red markup in offchain

ch1bo committed:
Add AssertionFailed logic error outcome and use it onDecrementTx

This, at least, results in log lines if this situation occurrs.

ch1bo committed:
Fix TxTraceSpec by modelling snapshot versions

By having versions in the ModelSnapshot, we can correctly test the
head validator with contest transactions that try to set wrong
versions (see removed TODO).

ch1bo committed:
Model: genPayment to any hydra party

ch1bo committed:
Revert changes on ObserveConfirmedTx

ch1bo committed:
Small refactor in genOpenActions

ch1bo committed:
Remove redundant e2e tests

All three were doing the same thing and the third one was also not a
"single hydra head" (as the category said).

ch1bo committed:
Simplify aliceAddress calculation

ch1bo committed:
Simplify decommit e2e scenario to only one decommit

ch1bo committed:
Further reduce complexity of decommit e2e test

By using zero-fee decommit tx, the amounts line up better.

ch1bo committed:
Rename a mutation enumeration to clarify

ch1bo commented:
Marking this ready to review now as basically only a rebase and PR descriptions are missing.

ch1bo committed:
Drop redundant test in TxSpec

The TxTraceSpec does test exactly that

ch1bo committed:
Regenerate plutus scripts after rebase

ch1bo committed:
Update genDecrement to also have sometimes empty commits

ch1bo committed:
Fix warnings

ch1bo committed:
Property test and implement splitUTxO differently

ch1bo committed:
Remove redundant Fanout mutation tests

One per validator code path is enough

ch1bo committed:
Start tuning TxTraceSpec

ch1bo committed:
Reduce discards of fanout

ch1bo committed:
Fix logs.yaml with forgotten OnDecrementTx

v0d1ch reviewed (approved):
Finally!

ch1bo committed:
Make model aware of outdated snapshots

ch1bo committed:
Remove/demote TODO and XXXs
